From 0888c6d8caa87adf0f5b727afeaaaca8f5289036 Mon Sep 17 00:00:00 2001
From: Shijie Qin <shijie.qin@nxp.com>
Date: Wed, 19 Sep 2018 16:37:59 +0800
Subject: [PATCH 4688/5242] MLK-19626: Verify wptr/rptr pointer is normal or
 not before update stream address When an abnormal
 condition causes the wptr/rptr pointer to be
 abnormal, cancel update stream address and report
 the error.

commit  9710d2d76a925a8564f1b47868a055bfb0422d9a from
https://source.codeaurora.org/external/imx/linux-imx.git

Signed-off-by: Shijie Qin <shijie.qin@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/vpu-decoder-b0/vpu_b0.c |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/mxc/vpu-decoder-b0/vpu_b0.c b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
index 13dd881..17e996a 100755
--- a/drivers/mxc/vpu-decoder-b0/vpu_b0.c
+++ b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
@@ -1374,6 +1374,14 @@ static int update_stream_addr(struct vpu_ctx *ctx, void *input_buffer, uint32_t
 
 	start = pStrBufDesc->start;
 	end = pStrBufDesc->end;
+	if (wptr < start || wptr > end) {
+		vpu_dbg(LVL_ERR, "%s(), wptr pointer cross-border\n", __func__);
+		return 0;
+	}
+	if (rptr < start || rptr > end) {
+		vpu_dbg(LVL_ERR, "%s(), rptr pointer cross-border\n", __func__);
+		return 0;
+	}
 	wptr_virt = (void *)ctx->stream_buffer_virt + wptr - start;
 
 	vpu_dbg(LVL_INFO, "update_stream_addr down\n");
-- 
1.7.9.5

